pkg/operation/operation.go

```diff
@@ -608,6 +610,16 @@ func (o *Operation) ComputeLokiHosts() []string {
 	}
 }
 
+// IsShootMonitoringEnabled returns true if shoot monitoring is enabled and shoot is not of purpose testing.
+func (o *Operation) IsShootMonitoringEnabled() bool {
```

Comment:
I'm wondering: should we also use this in `kubestatemetrics.go`?
Otherwise, `kube-state-metrics` will get deployed even if shoot monitoring is disabled.

IIRC, `kube-state-metrics` is scraped by the seed prometheus. Hence, I'm not sure if it's desired to keep `kube-state-metrics`. 
Do you want to keep it in your landscapes?

```diff
diff --git a/pkg/operation/botanist/kubestatemetrics.go b/pkg/operation/botanist/kubestatemetrics.go
index 313eddd92..5e540e114 100644
--- a/pkg/operation/botanist/kubestatemetrics.go
+++ b/pkg/operation/botanist/kubestatemetrics.go
@@ -17,7 +17,6 @@ package botanist
 import (
        "context"

-       gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
        "github.com/gardener/gardener/pkg/operation/botanist/component"
        "github.com/gardener/gardener/pkg/operation/botanist/component/kubestatemetrics"
        "github.com/gardener/gardener/pkg/utils/images"
@@ -45,7 +44,7 @@ func (b *Botanist) DefaultKubeStateMetrics() (kubestatemetrics.Interface, error)

 // DeployKubeStateMetrics deploys or destroys the kube-state-metrics to the shoot namespace in the seed.
 func (b *Botanist) DeployKubeStateMetrics(ctx context.Context) error {
-       if b.Shoot.Purpose == gardencorev1beta1.ShootPurposeTesting {
+       if !b.IsShootMonitoringEnabled() {
                return b.Shoot.Components.ControlPlane.KubeStateMetrics.Destroy(ctx)
        }

```

Reply:
> IIRC, `kube-state-metrics` is scraped by the seed prometheus. 

Ok, I double-checked this and it seems I didn't remember correctly.
`garden/prometheus` scrapes `garden/kube-state-metrics` including metrics for shoot namespaces.
`garden/seed-prometheus` scrapes `garden/kube-state-metrics` but drops metrics for shoot control plane namespaces.
`shoot--foo--bar/prometheus` scrapes `garden/kube-state-metrics` but only keeps metrics for the shoot control plane namespace (`shoot--foo--bar`).
`shoot--foo--bar/prometheus` scrapes `shoot--foo--bar/kube-state-metrics` which includes metrics for objects in the shoot cluster itself.

TL;DR: I think the `kube-state-metrics` for the shoot cluster itself should be disabled as well if the shoot monitoring is disabled.
I.e., you can simply apply the proposed patch from my last comment :)

Reply:
Sorry for the back and forth 🙊

Reply:
nice catch. Our older gardener version didn't had that yet `ಠ_ಠ`